strawberryfields/apps/train/parametrization.py

```diff
@@ -0,0 +1,363 @@
+        samples (array): an optional array of samples from :math:`A` used to speed up gradient
+            calculations #TODO: more info
+    """
+    def __init__(
+        self,
+        A: np.ndarray,
+        n_mean: float,
+        embedding,
+        threshold: bool,
+        samples: Optional[np.ndarray] = None,
+    ):
+        if not np.allclose(A, A.T):
+            raise ValueError("Input must be a NumPy array corresponding to a symmetric matrix")
+        self.A_init = A
+        self.embedding = embedding
+        self.threshold = threshold
+        self.A_init_scaled = rescale_adjacency(A, n_mean, threshold)
+        self.n_modes = len(A)
+        self._A_init_samples = None
+        if samples:
+            self._add_A_init_samples(samples)
+
+    def W(self, params: np.ndarray) -> np.ndarray:
+        """Calculate the diagonal matrix of weights :math:`W` that depends on the trainable
+        parameters :math:`\theta`.
+
+        **Example usage:**
+
+        >>> vgbs.W(params)
+        array([[0.95122942, 0.        , 0.        , 0.        ],
+               [0.        , 0.90483742, 0.        , 0.        ],
+               [0.        , 0.        , 0.98019867, 0.        ],
+               [0.        , 0.        , 0.        , 0.99004983]])
+
+        Args:
+            params (array): the trainable parameters :math:`\theta`
+
+        Returns:
+            array: the diagonal matrix of weights
+        """
+        return np.diag(self.embedding(params))
+
+    @staticmethod
+    def _WAW(A: np.ndarray, W: np.ndarray) -> np.ndarray:
+        """Calculate the :math:`_WAW` parametrization.
+
+        Args:
+            A (array): the adjacency matrix
+            W (array): the diagonal matrix of weights
+
+        Returns:
+            array: the :math:`_WAW` matrix
+        """
+        return W @ A @ W
+
+    def A(self, params: np.ndarray) -> np.ndarray:
+        """Calculate the trained adjacency matrix :math:`A(\theta)`.
+
+        **Example usage:**
+
+        >>> vgbs.A(params)
+        array([[0.        , 0.86070798, 0.93239382, 0.94176453],
+               [0.86070798, 0.        , 0.88692044, 0.89583414],
+               [0.93239382, 0.88692044, 0.        , 0.97044553],
+               [0.94176453, 0.89583414, 0.97044553, 0.        ]])
+
+        Args:
+            params (array): the trainable parameters :math:`\theta`
+
+        Returns:
+            array: the trained adjacency matrix
+        """
+        return self._WAW(self.A_init, self.W(params))
+
+    def _A_scaled(self, params: np.ndarray) -> np.ndarray:
+        """Calculate the trained scaled adjacency matrix :math:`A_{\rm scale}(\theta)`.
+
+        Args:
+            params (array): the trainable parameters :math:`\theta`
+
+        Returns:
+            array: the trained adjacency matrix
+        """
+        return self._WAW(self.A_init_scaled, self.W(params))
+
+    def _generate_samples(self, A: np.ndarray, n_samples: int, **kwargs) -> np.ndarray:
```

Comment:
I think it would be best to have a single `generate_samples()` function and simply specify the corresponding `A` matrix depending on the intended sampling. Having three separate sample functions, even if some are internal, seems unnecessary.

Reply:
Great, we now just have one `generate_samples()` that takes an adjacency matrix as input!